### ISSUE TITLE
Offer Expired Challenge: Based on Clock

### ISSUE THREAD

**Issue by agrawalarpit14** [juice-shop/juice-shop#847]
The flow of the challenge is like this:
Name: Clock Manipulation Challenge
Description: Order products on offer that expired in 2017
Tasks to be done: It would need the user to manipulate the clock of their machine
Flow: User goes to twitter and finds out a tweet about the offer which gives them the `/#/offer/<festive>` route, but when they try to access it they are shown `the offer has expired` but when they manipulate the clock it works out for them. They are able to order few products for free on this route.
Bug: After the offer was expired, the route was never removed.

**Comment by agrawalarpit14** [comment]
Hi @bkimminich @J12934 @CaptainFreak Please let me know if you find the challenge interesting, and other ideas and suggestions. Thank You
If you find it good I will start working on it?

**Comment by agrawalarpit14** [comment]
Hi @bkimminich I have tried to make the idea more concrete. The flow is as follows:
- The attacker finds about which route to go for the offer page in either of the following ways:
1. In browser, go to sources(in DevTools) and open the file `main.js` and find the route.
2. Go to twitter handle of juice-shop where you will find a tweet saying `On the occasion of Women's Day 2019, Juice shop brings you irresistible offers where few products are absolutely free, offer last till 12:00AM'.
- The attacker goes to the link but finds out the following response:
<img width="1680" alt="Screenshot 2019-03-11 at 2 06 04 PM" src="https://user-images.githubusercontent.com/35000671/54110242-e987e280-4406-11e9-9ca9-e31194c4de23.png">

- He then finds out that the route can be accessed by manipulation of local time as shown [here](https://drive.google.com/file/d/1eSJq3C8dwW3p3KdL7CxwGAmRHHPKF6tW/view?usp=sharing).
- The user tries the same and finds out that actually few products that exist on juice-shop were actually available for free.
- The user puts one of the products and complete the purchase to pass the challenge.

**NOTE**: The video shared above is not the final challenge, it's just for demonstration.
It would be very helpful if you could suggest some improvements.

**Comment by bkimminich** [comment]
Hm, I like the idea in general, but it's a discount calculated on the client side, right? As long is it does not compete with the coupon code concept too much, I'm fine with it. Maybe we could "blend it in" more into the coupon code field, say by checking for 10-char campaign-codes in the client (with your proposed time bug) and if no matching is found sending the code to the server-side for discount check like today? Examples:

* Code `WMNSDY2019` will be recognized as a campaign code and check if date is Friday, 8th March 2019 and if it is then give like 75% discount (or anything _less than 80%_ to not mess with the existing "Forged Coupon" challenge!) - :question: _How exactly did you plan to apply this discount type?_
* Code `mNYS#iw06w` will not be recognized as a campain code, so it goes into backend check, which will pass validity check, thus giving regular 40% off like today
* Code `1234567890` will not be recognized as campaign code, so it goes into backend check like today, which will fail and return the `Invalid coupon code` error as today

**Comment by agrawalarpit14** [comment]
Actually, I will go forward and implement the discount your way(through a coupon code)? Mine was a very naive way of implementing it, I was planning to add some kind of extra table in the database itself😛.

**Comment by supra08** [comment]
@agrawalarpit14, you can implement the same in a flash sale sort of thing too. 
* The user has very less amount of time while ordering in flash sales and the stocks are over after that.
* The flash sales also start from at a certain announced time.
* So the user would try to place the order of the product before the announced time by manipulating the clock and will also succeed since the stocks are still full.
This way the coupon area is not reused. And flash sales are common in retail sites for their esteemed products. What do you say, @bkimminich?

**Comment by bkimminich** [comment]
I love that flash sale idea! 👍😀

**Comment by bkimminich** [comment]
@supra08, you mean like https://www.packtpub.com/books/deal-of-the-day maybe, where one product is offered at a special price?

**Comment by supra08** [comment]
This is cool too, but what I meant was some that products, which are recently launched or are available in limited stocks are made available at a specific announced time, say 12:00. At this announced time, all the willing customers try to order as quickly as possible but the success rate is low because of the rush. In most occasions, the stocks are dead in 4-5 mins or so.
But seeing the context, the above idea, `deal of the day` fits in better.

**Comment by agrawalarpit14** [comment]
Hi @bkimminich @supra08 Thank you for your valuable input. The idea of flash sale sounds awesome to me. Actually, a flash sale is quite similar to what I have been originally planning(a route that will be restricted apart from a specific time period).
If you approve should I materialise it and make a different challenge out of it, as the present PR #859 is an entirely different challenge?

**Comment by bkimminich** [comment]
Yes, please go for it! We could make those then "On Sale Tier 1" and "Tier 2" challenges, but let's see which one turns out harder...

**Comment by bkimminich** [comment]
Oh, please just take a look at my review comments on your current PR before starting something new, @agrawalarpit14. Thanks! 👍

**Comment by agrawalarpit14** [comment]
Yeah sure. Thanks🙂

**Comment by supra08** [comment]
Would it be a good idea to use clock manipulation for two challenges?  The become almost the same.

**Comment by bkimminich** [comment]
Yeah, that's totally okay. We have like 4+ file access challenges. The important thing is to give each one a slightly different twist so they don't solve both a the same time. The _campaign discount_ challege @agrawalarpit14 wrote with my review comments applied would make it possible to solve it via clock manipulation or code analysis/request tampering. The _deal of the day_ could totally have one of both attack paths similar but should add a second one that is slightly different.

**Comment by bkimminich** [comment]
I'm closing this one as it's implemented already. If someone comes up with a _Deal of the Day_ scenario feel free to open a new issue so this one doesn't get too confusing. Thanks!
